src/app/codegen-data-model/CodegenDataModel_Write.cpp

```diff
@@ -0,0 +1,341 @@
+{
+    T workingValue;
+
+    if (isNullable)
+    {
+        typename DataModel::Nullable<T> nullableWorkingValue;
+        ReturnErrorOnFailure(decoder.Decode(nullableWorkingValue));
+
+        if (nullableWorkingValue.IsNull())
+        {
+            VerifyOrReturnError(out.size() >= sizeof(typename ENCODING::LengthType), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+            typename ENCODING::LengthType nullLength = ENCODING::kNullLength;
+            memcpy(out.data(), &nullLength, sizeof(nullLength));
+            return CHIP_NO_ERROR;
+        }
+
+        // continue encoding non-null value
+        workingValue = nullableWorkingValue.Value();
+    }
+    else
+    {
+        ReturnErrorOnFailure(decoder.Decode(workingValue));
+    }
+
+    typename ENCODING::LengthType len = static_cast<typename ENCODING::LengthType>(workingValue.size());
+    VerifyOrReturnError(out.size() >= sizeof(len) + len, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    memcpy(out.data(), &len, sizeof(len));
+    memcpy(out.data() + sizeof(len), workingValue.data(), workingValue.size());
+    return CHIP_NO_ERROR;
+}
+
+/// Encodes a numeric data value of type T from the given ember-encoded buffer `data`.
+///
+/// isNullable defines if the value of NULL is allowed to be encoded.
+template <typename T>
+CHIP_ERROR DecodeIntoEmberBuffer(AttributeValueDecoder & decoder, bool isNullable, MutableByteSpan out)
+{
+    if (isNullable)
+    {
+        using Traits = NumericAttributeTraits<T>;
+
+        DataModel::Nullable<typename Traits::WorkingType> workingValue;
+        ReturnErrorOnFailure(decoder.Decode(workingValue));
+
+        typename Traits::StorageType storageValue;
+        if (workingValue.IsNull())
+        {
+            Traits::SetNull(storageValue);
+        }
+        else
+        {
+            VerifyOrReturnError(Traits::CanRepresentValue(isNullable, *workingValue), CHIP_ERROR_INVALID_ARGUMENT);
+            Traits::WorkingToStorage(*workingValue, storageValue);
+        }
+
+        VerifyOrReturnError(out.size() >= sizeof(storageValue), CHIP_ERROR_INVALID_ARGUMENT);
+
+        const uint8_t * data = Traits::ToAttributeStoreRepresentation(storageValue);
+        memcpy(out.data(), data, sizeof(storageValue));
+    }
+    else
+    {
+        using Traits = NumericAttributeTraits<T>;
+
+        typename Traits::WorkingType workingValue;
+        ReturnErrorOnFailure(decoder.Decode(workingValue));
+
+        typename Traits::StorageType storageValue;
+        Traits::WorkingToStorage(workingValue, storageValue);
+
+        VerifyOrReturnError(out.size() >= sizeof(storageValue), CHIP_ERROR_INVALID_ARGUMENT);
+
+        // This guards against trying to encode something that overlaps nullable, for example
+        // Nullable<uint8_t>(0xFF) is not representable because 0xFF is the encoding of NULL in ember
+        VerifyOrReturnError(Traits::CanRepresentValue(isNullable, workingValue), CHIP_ERROR_INVALID_ARGUMENT);
```

Comment:
- IsNullable is always false here
- It is always valid to store 0xFF in a non-nullable field. This line should be removed.

Reply:
This is generic for types that are non-string-like. Specific example is the odd-sized integers. In that case say a 32-bit integer is used as "working value" however the "storage value" could be a 3-byte value.

In that case, CanRepresentValue(0x1234abcd) will return false even if non-nullable because we can only represent 0 ... 0xFFFFFF.

Reply:
I will update comments - I think I got a copy and paste.

For isNullable it is true that on each branch I know what it is (true or false), however I figured booleans are hard to read so I would rather have `CanRepresentValue(isNullable, workingValue)` then `CanRepresentValue(false /* isNullable */, workingValue)` since I then have the same functionality, but longer code and more likely to mis-use by copy & paste.